libbpf-rs/src/object.rs

```diff
@@ -300,65 +435,40 @@ impl Object {
     /// undefined.
     ///
     /// It is not safe to manipulate `ptr` after this operation.
-    pub unsafe fn from_ptr(ptr: *mut libbpf_sys::bpf_object) -> Self {
+    pub unsafe fn from_ptr(ptr: *mut libbpf_sys::bpf_object) -> Result<Self> {
         Self::new(ptr)
     }
 
-    pub fn map<T: AsRef<str>>(&mut self, name: T) -> Result<Option<&mut Map>> {
-        if self.maps.contains_key(name.as_ref()) {
-            Ok(self.maps.get_mut(name.as_ref()))
-        } else if let Some(ptr) = find_map_in_object(self.ptr, name.as_ref())? {
-            let owned_name = name.as_ref().to_owned();
-            let fd = unsafe { libbpf_sys::bpf_map__fd(ptr) };
-            if fd < 0 {
-                Err(Error::System(errno::errno()))
-            } else {
-                // bpf_map__def can return null but only if it's passed a null. Object::map
-                // already error checks that condition for us.
-                let def = unsafe { ptr::read(libbpf_sys::bpf_map__def(ptr)) };
-
-                self.maps.insert(
-                    owned_name.clone(),
-                    Map::new(fd, owned_name, def.type_, def.key_size, def.value_size, ptr),
-                );
-
-                Ok(self.maps.get_mut(name.as_ref()))
-            }
-        } else {
-            Ok(None)
-        }
+    /// Get a mutable reference to `Map` with the name `name`, if one exists.
+    pub fn map<T: AsRef<str>>(&mut self, name: T) -> Option<&mut Map> {
+        self.maps.get_mut(name.as_ref())
     }
 
-    /// Same as [`Object::map`] except will panic if `Err` or `None` is encountered.
-    pub fn map_unwrap<T: AsRef<str>>(&mut self, name: T) -> &mut Map {
-        self.map(name).unwrap().unwrap()
+    /// Get an iterator over references to all `Map`s.
+    /// Note that this will include automatically generated .data, .rodata, .bss, and
+    /// .kconfig maps. You may wish to filter these out depending on your use case.
+    pub fn maps_iter_mut(&mut self) -> impl Iterator<Item = &mut Map> {
```

Comment:
`map()` and `prog()` already return mutable references without `mut` in their name. Maybe rename to `maps_iter()`? However, the API guidelines say otherwise: https://rust-lang.github.io/api-guidelines/naming.html#methods-on-collections-that-produce-iterators-follow-iter-iter_mut-into_iter-c-iter . Don't have a strong opinion here, wondering what others think

Reply:
I was a bit torn on this for this reason. I do think we should try and be as consistent as possible with the API guidelines. When I see something like `maps_iter` I think I am going to get something immutable back.

Reply:
Something else to consider would be whether it's worth supporting the ability to grab immutable references to maps.

Reply:
For maps specifically, it might be worthwhile to have an immutable reference because there are some useful functions that can be used with just a reference. For programs though, it seems much less useful to get an immutable reference. Just my two cents.

I would be okay with the explicit naming, but if there isn't explicit naming I think the library should probably just try to be consistent.

Reply:
If we do end up adding immutable versions, perhaps that should be in a separate PR (possibly with changes to generated skeleton file as well?). I do agree that immutable map references would be useful for many cases.

Reply:
For now, I'll revert my current naming convention back to `maps_iter` and `progs_iter` and then we can address immutability in a future PR.